Test runner

arcusfelis opened the pull request:
This PR adds `test-runner.sh` script.

Proposed changes include:
* updated docs https://github.com/esl/MongooseIM/blob/test-runner/doc/developers-guide/Testing-MongooseIM.md
* Script that prepares env variables for `travis-test.sh`.
* Argument autocompletion script.

mongoose-im commented:
[5100.1](https://travis-ci.org/esl/MongooseIM/jobs/402243111) / Erlang 19.3 / small_tests / ae029ef47376e8147ab2ed82d4ff2d7623bcaf40
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5100/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5100/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-10_14.40.29/lib.mongooseim.logs/run.2018-07-10_14.40.30/suite.log.html)

mongoose-im commented:
[5102.1](https://travis-ci.org/esl/MongooseIM/jobs/402247566) / Erlang 19.3 / small_tests / 7f63dbf72e7b48f747054a1e44efd679803b1697
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5102/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5102/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-10_15.30.13/lib.mongooseim.logs/run.2018-07-10_15.30.14/suite.log.html)

michalwski commented on doc/developers-guide/Testing-MongooseIM.md:
````diff
@@ -1,3 +1,54 @@
+./tools/test-runner.sh --db redis --preset internal_mnesia
+```
+
+The command runs both big feature and small unit tests.
````
Maybe:
The command runs both big (feature) and small (unit) tests.

michalwski commented on tools/test_runner/show_reports.sh:
```diff
@@ -0,0 +1,34 @@
+
+    echo "Running an HTTP server for $ABS_DIRECTORY"
+    python -m SimpleHTTPServer "$PORT" >> http_server.log 2>> http_server.log &
+    echo "Open http://localhost:$PORT in your browser"
```
1. There is no need to for the python's HTTP server, we can open the `${ABS_DIRECTORY}/index.html` or `${ABS_DIRECTORY}/all_runs.html`.
2. Instead of only printing `Open sth in your browser` we can simply call
    ```bash
    open  ${ABS_DIRECTORY}/index.html
    ```
   and this will open the report in the default browser. I would still leave the URL printing so that the user sees in the console what was opened.

arcusfelis replied:
not really, chrome does not load xslt stylesheets without a browser.

"open" would not work on a remote shell or with a separate user for code.

`open` is also platform-specific.

Yeah, I can try to call `open` and `xdg-open`, but it would not always work.

michalwski replied:
Ok, let's leave it as is. Thanks for the explanation.

arcusfelis commented on doc/developers-guide/Testing-MongooseIM.md:
````diff
@@ -1,3 +1,54 @@
+./tools/test-runner.sh --examples
+```
+
+## Test runner competion
````
You win, if tests pass.

mongoose-im commented:
[5112.1](https://travis-ci.org/esl/MongooseIM/jobs/402657247) / Erlang 19.3 / small_tests / 4d8c069a50e30ec873f586f97649411dba374ad5
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5112/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5112/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-11_13.38.40/lib.mongooseim.logs/run.2018-07-11_13.38.41/suite.log.html)

erszcz commented on big_tests/run_common_test.erl:
```diff
@@ -1,18 +1,18 @@
 %%
 %% Example with two nodes:
-%% DEV_NODES="mim1 mim2" TEST_HOSTS="mim mim2" ./tools/travis-test.sh -e false -c false -s false -p odbc_mssql_mnesia
+%% DEV_NODES="mim1 mim2" TEST_HOSTS="mim mim2" ./tools/travis-test.sh -c false -s false -p odbc_mssql_mnesia
```
This does not seem to be consistent with modified line 2 above. Here we have `TEST_HOSTS="mim mim2"`, but above there's a change from `TEST_HOSTS="mim"` to `TEST_HOSTS="mim1"` - shouldn't we follow the example here, too?

arcusfelis replied:
fixed

michalwski commented:
The code looks ok. I don't like the fact that there is more 1k new lines added for this, the only other option would be to extract most of the scripts to another repo. This may also be a not good idea.

I'm going to try this script on my Mac to see if this works for me.

mongoose-im commented:
[5122.1](https://travis-ci.org/esl/MongooseIM/jobs/403052198) / Erlang 19.3 / small_tests / 560dcaa84c68948676cd16553d18988397c554d0
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5122/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5122/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-12_10.32.37/lib.mongooseim.logs/run.2018-07-12_10.32.38/suite.log.html)

arcusfelis commented:
@michalwski  
1. `--db -- THE_REST_OF_ARGUMENTS` to skip db-setup.
2. `--dev-nodes --` to rerun big-tests.

mongoose-im commented:
[5123.1](https://travis-ci.org/esl/MongooseIM/jobs/403061771) / Erlang 19.3 / small_tests / d3e9ef415aa0ca4caadf7b08e7526756ba55c291
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5123/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5123/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-12_11.07.25/lib.mongooseim.logs/run.2018-07-12_11.07.26/suite.log.html)

goddammit commented on doc/developers-guide/Testing-MongooseIM.md:
```diff
@@ -1,3 +1,54 @@
+
+Test runner script is used to compile MongooseIM and run tests.
+
+Help command prints a list of supported options.
```
The help command prints a list of supported options.

goddammit commented on doc/developers-guide/Testing-MongooseIM.md:
```diff
@@ -1,3 +1,54 @@
+# Test runner
+
+Test runner script is used to compile MongooseIM and run tests.
```
The test runner script is used to compile MongooseIM and run tests.

goddammit commented on doc/developers-guide/Testing-MongooseIM.md:
```diff
@@ -15,8 +66,40 @@ Detailed test results in a nice HTML format are saved in
+./tools/test-runner.sh --skip-small-tests --db mysql pgsql --preset mysql_mnesia pgsql_mnesia -- mam:odbc_muc_light
+
+# Runs rdbms_SUITE with MSSQL
+# Inits single MongooseIM node (works for some tests only)
```
Inits a single MongooseIM node (works for some tests only)

goddammit commented on tools/setup-redis.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+docker rm -f mongooseim-redis || echo "Skip removing previous container"
```
Skip removing the previous container

goddammit commented on tools/test-runner-complete.sh:
```diff
@@ -0,0 +1,161 @@
+_run_all_tests() {
+  printf "%s\n" "${COMP_WORDS[@]}" > /tmp/test-runner-last-competion
+  # Make COMP_WORDS, without using colon as a breaker
+  # To see all breakes, check COMP_WORDBREAKS variable
```
breakes or breakers?

goddammit commented on tools/test-runner-complete.sh:
```diff
@@ -0,0 +1,161 @@
+  cur=${COMP_WORDS[$COMP_CWORD]}
+  opt=${COMP_WORDS[$pos]}
+
+  # If current is option - we don't care about option expantion
```
If current is an option - we don't care about option expansion ?

goddammit commented on tools/test-runner.sh:
```diff
@@ -0,0 +1,516 @@
+# - VERBOSE
+# - STOP_NODES
+
+# Just compile big tests example (not really):
```
?

arcusfelis replied:
meh, impossible feature

arcusfelis commented:
@michalwski 

> It's not clear and intuitive how to skip setup db. Of course this can be done by --db "" (probably). I think it'd be more intuitive and clear if it was possible to skip db setup by using option like --skip-db-set (even if there is --db option set).

I've added `skip-setup-db`.

>  It's not clear how to rerun big tests (or a suite from big tests without rebuilding and restarting MongooseIM nodes). I think it'd be good to describe how to test, change (code or test), test loop efficiently with test runner. Keep in mind that sometimes we want to change only tests and re-run it quickly without changing any code.

I've added `--rerun-big-tests` option.

```erlang
./tools/test-runner.sh --rerun-big-tests -- mam:odbc_simple_muc_all
```

> I'm not sure it makes much sense to start all db backends if there was no --db option set. Especially if we want to run preset internal_mnesia. Maybe we could have a mapping where we specify what dbs are required for a specific preset and if there is no --db option passed we start only these required? This point is only observation/feedback, not necessarily sth to change immediately.

OK, adding

mongoose-im commented:
[5132.1](https://travis-ci.org/esl/MongooseIM/jobs/403153531) / Erlang 19.3 / small_tests / 03e7b78ae225881c9eadc2c3aa9106f28f61f64c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5132/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5132/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-12_15.20.36/lib.mongooseim.logs/run.2018-07-12_15.20.37/suite.log.html)

mongoose-im commented:
[5136.1](https://travis-ci.org/esl/MongooseIM/jobs/403300242) / Erlang 19.3 / small_tests / b66b1dc0b316f84b0b9372d13551a5ed5c6155dc
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5136/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5136/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-12_21.07.00/lib.mongooseim.logs/run.2018-07-12_21.07.01/suite.log.html)

michalwski commented on big_tests/test.config:
```diff
@@ -181,15 +181,19 @@
     {internal_mnesia,
-     [{sm_backend, "{mnesia, []}"},
+     %% dbs variable is used by ./tools/test_runner/presets_to_dbs.sh script
+     [{dbs, [redis]},
```
I would say we don't need `redis` for the `internal_mnesia` preset. Do we?

arcusfelis commented:
We need redis for global distribution stuff.

On Fri, 13 Jul 2018, 09:00 Michał Piotrowski, <notifications@github.com>
wrote:

> *@michalwski* approved this pull request.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/esl/MongooseIM/pull/1973#pullrequestreview-136924242>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnDNDplTq8jy_vpOnCvbZXDlYcV7cOPks5uGEV0gaJpZM4VJhrU>
> .
>

michalwski commented:
> We need redis for global distribution stuff

Thanks @arcusfelis for the explanation, I completely forgot about it.

mongoose-im commented:
[5159.1](https://travis-ci.org/esl/MongooseIM/jobs/404443308) / Erlang 19.3 / small_tests / 09a22da551b4216a5f4d17ea5e968a438eb7a551
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1973/5159/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1973/5159/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-16_13.33.02/lib.mongooseim.logs/run.2018-07-16_13.33.03/suite.log.html)
